app/src/main/kotlin/com/mapbox/vision/teaser/SafetyFragment.kt

```diff
@@ -0,0 +1,238 @@
+                if (currentDangerLevel != collision.dangerLevel) {
+                    soundsPlayer.stop()
+                    when (collision.dangerLevel) {
+                        CollisionDangerLevel.None -> Unit
+                        CollisionDangerLevel.Warning -> soundsPlayer.playWarning()
+                        CollisionDangerLevel.Critical -> soundsPlayer.playCritical()
+                    }
+                    currentDangerLevel = collision.dangerLevel
+                }
+            }
+
+            fun showCollision(collision: CollisionObject) {
+                distance_to_car_label.show()
+                safety_mode.show()
+                distance_to_car_label.text =
+                        distanceFormatter.formatDistance(collision.`object`.position.y)
+
+                when (currentDangerLevel) {
+                    CollisionDangerLevel.None -> safety_mode.clean()
+                    CollisionDangerLevel.Warning -> safety_mode.drawWarnings(collisions)
+                    CollisionDangerLevel.Critical -> safety_mode.drawCritical()
+                }
+            }
+
+            fun hideCollision() {
+                soundsPlayer.stop()
+                currentDangerLevel = CollisionDangerLevel.None
+                distance_to_car_label.hide()
+                safety_mode.hide()
+            }
+
+            fun calibrationReady() {
+                distance_to_car_label.show()
+                safety_mode.show()
+                calibration_progress.hide()
+
+                val collision = collisions.firstOrNull { it.`object`.objectClass == DetectionClass.Car }
+                if (collision == null) {
+                    hideCollision()
+                } else {
+                    updateCollisionDangerLevelIfChanged(collision)
+                    showCollision(collision)
+                }
+            }
+
+            fun calibrationInProgress() {
+                distance_to_car_label.hide()
+                safety_mode.hide()
+                calibration_progress.show()
+                val progress = (calibrationProgress * 100).toInt()
+                calibration_progress.text = getString(R.string.calibration_progress, progress)
+            }
+
+            runOnUiThread {
+                if (this@SafetyFragment.isResumed) {
+                    if (calibrationProgress == CALIBRATION_READY_VALUE) {
+                        calibrationReady()
+                    } else {
+                        calibrationInProgress()
+                    }
+                }
+            }
+        }
+
+        val speedLimitTranslation by lazy {
+            resources.getDimension(R.dimen.speed_limit_translation)
+        }
+
+        override fun onRoadRestrictionsUpdated(roadRestrictions: RoadRestrictions) {
+            runOnUiThread {
+                if (this@SafetyFragment.isResumed) {
+                    fun getImageResource() = signResources.getSpeedSignResource(
```

Comment:
I'd moved fun definitions out of `runOnUiThread`

Reply:
fixed